hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+//! [ascent] datalog implementation of analysis.
+//! Since ascent-(macro-)generated code generates a bunch of warnings,
+//! keep code in here to a minimum.
+#![allow(
+    clippy::clone_on_copy,
+    clippy::unused_enumerate_index,
+    clippy::collapsible_if
+)]
+
+use ascent::lattice::{BoundedLattice, Lattice};
+use itertools::zip_eq;
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::ops::{Index, IndexMut};
+
+use hugr_core::extension::prelude::{MakeTuple, UnpackTuple};
+use hugr_core::ops::{OpTrait, OpType};
+use hugr_core::{HugrView, IncomingPort, Node, OutgoingPort, PortIndex as _};
+
+use super::{AbstractValue, DFContext, PartialValue};
+
+type PV<V> = PartialValue<V>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum IO {
+    Input,
+    Output,
+}
+
+ascent::ascent! {
+    pub(super) struct AscentProgram<V: AbstractValue, C: DFContext<V>>;
+    relation context(C);
```

Comment:
There are a lot of copies of `C`, for every element of every relation, and then again in the indexes. Even when `C` is just an `Arc` that presumably adds to be quite substantial. This could be addressed by using `ascent_run` in a function where the context is in scope. We could then also avoid the hacky `Hash` implementations on contexts.

Reply:
Yup. My bad here - I never looked at the top-level doc for ascent to see `ascent_run`. Thanks :-), that is about 100* better. (Lead to moving a few things around, and separating DFContext from HugrView.)